doc/tutorial/tutorial.md

```diff
@@ -815,6 +842,23 @@ pub fn string_slice(x: &String, from: &u64, to: &u64) -> String {
 DDlog will automatically pickup this file and inline its contents in the
 generated `lib.rs`.
 
+#### Functions with side effects
+
+Functions implemented completely in DDlog are pure (side-effect-free) computations.
+It is however possible to declare extern functions with side effects.  The DDlog
+compiler needs to know about these side effects, as they may interfere with its
+optimizations.  The programmer is responsible for labeling such functions with
+the `#[has_side_effects]` attribute, e.g., the following function is defined in
+the `log.dl` library:
+
+```
+#[has_side_effects]
+extern function log(module: module_t, level: log_level_t, msg: string): ()
+```
+
+The compiler automatically infers these annotations for non-extern functions, so
```

Comment:
This seems inconsistent, before you said that non-extern functions cannot have side-effects. So what annotations are inferred?

Reply:
non-extern functions can invoke extern functions with side effects.  I will clarify.